### DISCUSSION TITLE
Evaluate value channel content, stop if channel value <= 3

### DISCUSSION THREAD

**Discussion by hkaspersen** [nextflow-io/nextflow#3769] [Q&A] [answered]
I have a file that is generated in a process that filters the data based on a parameter given by the user.
Thus, this filtering may therefore remove too many rows in the data, making it unusable for downstream analyses.

I want to make sure that the pipeline stops when it sees that the number of rows in the resulting file is <= 3. However, I have not managed to implement this yet, and I am unsure how.

I'd like it to work something similar to this:

```
nlines = filtering_process.out.filtered_data_ch
                .splitText()
                .count()

if (nlines <= 3) {
                exit 1, "Too few data points remaining after filtering"
}
```
Not sure if this is the best way to handle these things though, I appreciate any help!

**Comment by mribeirodantas** [comment, marked as answer]
You can have a `when` block in your process whose conditional expression is based on opening the file (that I believe is an input to this process) and checking how many rows it contains. This way, if the file contains `<N` rows, the process won't instantiate a task.

**Reply by hkaspersen** [reply]
Thanks for the suggestion! Is there a way to combine this with a message somehow?

**Reply by mribeirodantas** [reply]
Well, depending on what you want to do in your script block, you can have `exec` instead and throw an exception depending on the condition (you wouldn't use `when` in this situation).

**Reply by hkaspersen** [reply]
Is there any documentation for these kind of things I can read? I cannot seem to find any mentions of this in the nextflow documentation.

**Reply by mribeirodantas** [reply]
Throwing an exception? Here is one example: https://github.com/nf-core/eager/blob/bb32ae3b0110b9a26b791f73a5324828d849271a/main.nf#L3340

It's basically Groovy, so you should check for it in Groovy documentation.

**Reply by hkaspersen** [reply]
I am currently trying to make this work:

```
n_lines = PROCESS_NAME.out.filter_ch
                .map{ it -> it.countLines() }

process NAME {
        input:
        path(report)
        val(clustering_method)
        val(n_lines)

        script:
        if (n_lines < 4)
            throw Exception("Too few samples passed filtering, please increase --max_missing")
        else
            """
            Rscript script.R $clustering_method
            """
}

NAME(report_ch,
           clust_ch,
           n_lines)
```

However this is giving me the following error message:

```
Error executing process > 'NAME (1)'

Caused by:
  No signature of method: nextflow.script.ScriptBinding.GroovyException() is applicable for argument types: () values: []

Source block:
  if (n_lines < 4)
            throw Exception("Too few genomes passed filtering, please adjust --max_missing")
        else
  ======    """
            Rscript script.R $clustering_method
            """
```

Tried to google this but can't seem to find a solution for this error. The process works fine when the number of lines is >= 4.

**Reply by mribeirodantas** [reply]
You missed the `new` keyword before the keyword `exception`. It should be:

```
...

throw new Exception("Too few..
...
```

**Reply by hkaspersen** [reply]
That worked, thanks!

**Reply by mribeirodantas** [reply]
🥳
